CSS overflow module (#26425)
mdn/content: files/en-us/web/css/css_overflow/index.md

{{CSSRef}}

The **CSS overflow** module properties enable you to handle scrollable overflow in visual media.

Overflow happens when the content in an element box extends past one or more of the box's edges. **Scrollable overflow** is the content that appears outside the element box for which you might want to add a scrolling mechanism. CSS overflow properties enable you to control what happens when content overflows an element box.

Painting effects that overflow the content but do not participate in the CSS box model do not affect layout. This type of overflow is also known as {{Glossary("ink overflow")}}. Examples of ink overflows include box shadows, border images, text decoration, overhanging glyphs, and outlines. Ink overflows do not extend the scrollable overflow region.

## Overflow in action

Use the following interactive example to see the effects of various `overflow` property values on the content overflow and scrollbars in the adjacent fixed-size box.

> The example also includes options to change the values for the `overflow-clip-margin` and `width` properties, as well as to programmatically scroll the content if the overflow property creates a {{Glossary("scroll container")}}. Select `overflow: clip` and see the effect of different `overflow-clip-margin` values. Select `overflow: hidden` or `overflow: scroll` to check out different `ScrollLeft` and `ScrollTop` slider settings. 

Author: ```suggestion
The example also includes options to change the values for the `overflow-clip-margin` and `width` properties, as well as to programmatically scroll the content if the overflow property creates a {{Glossary("scroll container")}}. Select `overflow: clip` and see the effect of different `overflow-clip-margin` values. Select `overflow: hidden` or `overflow: scroll` to check out different `ScrollLeft` and `ScrollTop` slider settings.
```

{{CSSRef}}

The **CSS Overflow** module contains the features of CSS relating to scrollable overflow handling in visual media.

CSS overflow happens when the content of a box extends past one or more of the box's edges. **Scrollable overflow** is content that appears outside of the box for which scrolling mechanisms need to be provided. The overflow properties are how we can control what happens when content overflows a box.

Painting effects that overflow content but do not participate in the CSS box model, also known as {{Glossary("ink overflow")}}, do not affect layout. Ink overflows, such as box shadows, border images, text decoration, overhanging glyphs, outlines, etc., do not extend the scrollable overflow region.

## Overflow in action

> The following interactive example shows the effects of changing the value of the `overflow` property changes overflow of a fixed-size box. The example also includes options to change the `width` and `overflow-clip-margin` property values, as well as to programmatically scroll the content if the overflow property creates a {{Glossary("scroll container")}}.

Reviewer: ```suggestion
Use the following interactive example to see the effects of various `overflow` property values on the content overflow and scrollbars in the adjacent fixed-size box.

The example also includes options to change the values for the `overflow-clip-margin` and `width` properties, as well as to programmatically scroll the content if the overflow property creates a {{Glossary("scroll container")}}. Select `overflow: clip` and see the effect of different `overflow-clip-margin` values. Select `overflow: hidden` or `overflow: scroll` to check out different `ScrollLeft` and `ScrollTop` slider settings. 
```
- This example is really awesome!! Thanks for putting this together 🙌 
- When I select `overflow: overlay`, the heading in the content box reads "overlay (or clip if not supported):". Should `clip` be `auto`?

{{EmbedGHLiveSample("css-examples/modules/overflow.html", '100%', 320)}}

> A link is included in the preformatted lyrics to demonstrate the effects of keyboard focus on overflow and scroll behaviors. Try tabbing to the link or programmatically scrolling the content: the content will only scroll if the enumerated `<overflow>` value created a scroll container.

Reviewer: ```suggestion
A link is included in content box above to demonstrate the effects of keyboard focus on overflow and scroll behaviors. Try tabbing to the link or programmatically scrolling the content: the content will scroll only if the enumerated `<overflow>` value created a scroll container.
```
- Tabbing to the link is supported in chrome but not in Firefox
- BTW, in chrome when `clip` is selected:
![Screenshot 2023-04-26 at 7 12 56 PM](https://user-images.githubusercontent.com/23019147/234722284-d1f26d31-9c8e-498c-9f77-105a0b82aa72.png)
In Firefox, with `clip` selected, programmatic options are not disabled:
![Screenshot 2023-04-26 at 7 13 13 PM](https://user-images.githubusercontent.com/23019147/234722351-a5538d9d-3ad8-44c8-bc02-3572ff7cf35e.png)

Author: the css uses :has() to grey out useless CSS in supporting browsers. That feature is an enhancement, not a requirement.

## Reference

### CSS properties

> - {{CSSxRef("overflow-block")}}
> - {{CSSxRef("overflow-inline")}}
> - {{CSSxRef("overflow-x")}}
> - {{CSSxRef("overflow-y")}}
> - {{CSSxRef("overflow")}} shorthand
> - {{CSSxRef("overflow-clip-margin")}}
> - {{CSSxRef("scroll-behavior")}}
> - {{CSSxRef("scrollbar-gutter")}}
> - {{CSSxRef("text-overflow")}}

Reviewer: alphabetical

```suggestion
- {{CSSxRef("overflow")}} shorthand
- {{CSSxRef("overflow-block")}}
- {{CSSxRef("overflow-clip-margin")}}
- {{CSSxRef("overflow-inline")}}
- {{CSSxRef("overflow-x")}}
- {{CSSxRef("overflow-y")}}
- {{CSSxRef("scroll-behavior")}}
- {{CSSxRef("scrollbar-gutter")}}
- {{CSSxRef("text-overflow")}}
```

> - {{CSSxRef("block-ellipsis")}} (formerly {{CSSxRef("block-overflow")}}) {{experimental_inline}}

Reviewer 2: There is an `overflow-block` in Firefox (https://developer.mozilla.org/en-US/docs/Web/CSS/overflow-block ). Is it the same as that `block-overflow`? If not, and nobody has implemented `block-overflow`, we shouldn't even mention it.

Reviewer 3: It seems from the spec [history](https://w3c.github.io/csswg-drafts/css-overflow/#changes-since-2018-07-31) that `block-overflow` was renamed to `block-ellipsis` to avoid the very confusion.

> - {{CSSxRef("block-ellipsis")}} (formerly {{CSSxRef("block-overflow")}}) {{experimental_inline}}

Author: ```suggestion
- {{CSSxRef("block-ellipsis")}}
```

- {{CSSxRef("continue")}} {{experimental_inline}}
- {{CSSxRef("line-clamp")}} {{experimental_inline}}

> - {{CSSxRef("max-lines")}} {{experimental_inline}}

Reviewer: Should these be part of the "Related concepts" list?

Author: -webkit-line-clamp
block-ellipsis
continue
line-clamp
max-lines
are all in CSS-OVERFLOW-4

- {{CSSxRef("-webkit-line-clamp")}} {{non-standard_inline}}

> ### Data types

Reviewer: How about this? 🙂 

```suggestion
### Enumerated values
```

Author: i want to keep it the same title as other landing pages.

Reviewer: We can revisit this title if and when we do have the "Enumerated value types" category

- [`<overflow>`](/en-US/docs/Web/CSS/overflow_value) enumerated values

> ### Glossary
> 
> - {{Glossary("Scroll container")}}
> - {{Glossary("Ink overflow")}}
> 

Reviewer: ```suggestion
```

These need to go in the "Related concepts" section. Thats what we did in the [colors](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Colors#related_concepts) and [animations](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Animations#related_concepts) modules.

## Guides

- [Overflowing content](/en-US/docs/Learn/CSS/Building_blocks/Overflowing_content)
  - : CSS building block: learn what overflow is and how to manage it.
- {{Experimental_Inline}} [Creating a named scroll timeline](/en-US/docs/Web/CSS/scroll-timeline-name#creating_a_named_scroll_timeline)
  - : The CSS scroll timeline {{cssxref('scroll-timeline-name')}} and {{cssxref('scroll-timeline-axis')}} properties, along with the {{cssxref('scroll-timeline')}} shorthand, create animations tied to the scroll offset of a scroll container.

## Related concepts

> - {{CSSxRef("scrollbar-width")}}
> - {{CSSxRef("scrollbar-color")}}

Reviewer: ```suggestion
- {{CSSxRef("scrollbar-width")}} CSS property
- {{CSSxRef("scrollbar-color")}} CSS property
```

- {{CSSxRef("scroll-behavior")}} CSS property
- {{CSSxRef("scrollbar-gutter")}} CSS property
- {{cssxref("scroll-margin")}} CSS shorthand property
- {{cssxref("scroll-padding")}} CSS shorthand property
- {{cssxref("scroll-snap-align")}} CSS property
- {{cssxref("scroll-snap-stop")}} CSS property
- {{cssxref("scroll-snap-type")}} CSS property
- {{cssxref("text-overflow")}} CSS property
- {{CSSxRef("::-webkit-scrollbar")}} pseudo-element
- [`scrollbar`](/en-US/docs/Web/Accessibility/ARIA/Roles/scrollbar_role) ARIA role
- Element {{domxref("Element.scroll", "scroll()")}} method
- Element {{domxref("Element.scrollBy", "scrollBy()")}} method
- Element {{domxref("Element.scrollIntoView", "scrollIntoView()")}} method
- Element {{domxref("Element.scrollTo", "scrollTo()")}} method
- Element {{domxref("Element.scrollTop", "scrollTop")}} property
- Element {{domxref("Element.scrollLeft", "scrollLeft")}} property
- Element {{domxref("Element.scrollWidth", "scrollWidth")}} property
- Element {{domxref("Element.scrollHeight", "scrollHeight")}} property
- Document {{domxref("Document.scroll_event", "scroll")}} event

> 

Reviewer: ```suggestion
- {{Glossary("Scroll container")}} glossary term
- {{Glossary("Ink overflow")}} glossary term

```

## Specifications

{{Specifications}}

## See also

> - [CSS scrollbars styling module](/en-US/docs/Web/CSS/CSS_scrollbars)

Reviewer 2: ```suggestion
- [CSS Scrollbars](/en-US/docs/Web/CSS/CSS_Scrollbars)
```

Reviewer 2: I wonder if we should start with one PR that starts by fixing all module page titles, filenames, and slugs (all modules, in all of mdn/content), making it easier to link to them in a consistent way everywhere. Because for the moment, renaming happens in one PR, fixing in others, and filenames are not fixed, making reviewing these PRs difficult. (And leaving a trail of flaws on MDN). Then it will be easier to focus on the content. cc/ @dipikabh

Author: The name of the module is literally "CSS scrollbars styling module"

Reviewer: In this particular case, yes, the module name is "CSS scrollbars styling module". It is the slug that we're trying to keep as is and that's a mismatch with the page title.

I have been wondering the same, @teoli2003. Thanks for bringing this up. There's also the problem of casing in the slugs. They [should follow sentence casing](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Writing_style_guide#slugs) and they're inconsistent at the moment.

I can open the PR for the CSS part because that's what I am more familiar with (page titles, filenames, slugs). Are there modules or such changes required in other parts?

Also to confirm, all slug changes would require `yarn content move <from-slug> <to-slug>` to be run, right?

> - [CSS scrollbars styling module](/en-US/docs/Web/CSS/CSS_scrollbars)
> - [CSS scroll snap module](/en-US/docs/Web/CSS/CSS_Scroll_Snap)
> - [CSSOM view module](/en-US/docs/Web/CSS/CSSOM_View)
> - How to [Debug scrollable overflow](https://firefox-source-docs.mozilla.org/devtools-user/page_inspector/how_to/debug_scrollable_overflow/index.html).

Reviewer: ```suggestion
- [CSS scrollbars styling](/en-US/docs/Web/CSS/CSS_scrollbars) module
- [CSS scroll snap](/en-US/docs/Web/CSS/CSS_Scroll_Snap) module
- [CSSOM view](/en-US/docs/Web/CSS/CSSOM_View) module
- How to [debug scrollable overflow](https://firefox-source-docs.mozilla.org/devtools-user/page_inspector/how_to/debug_scrollable_overflow/index.html)
```

## Overflow in action

The following interactive example shows the effects of changing the value of the `overflow` property changes overflow of a fixed-size box. The example also includes options to change the `width` and `overflow-clip-margin` property values, as well as to programmatically scroll the content if the overflow property creates a {{Glossary("scroll container")}}.

{{EmbedGHLiveSample("css-examples/modules/overflow.html", '100%', 320)}}

A link is included in the preformatted lyrics to demonstrate the effects of keyboard focus on overflow and scroll behaviors. Try tabbing to the link or programmatically scrolling the content: the content will only scroll if the enumerated `<overflow>` value created a scroll container.

## Reference

### CSS properties

- {{CSSxRef("overflow-block")}}
- {{CSSxRef("overflow-inline")}}
- {{CSSxRef("overflow-x")}}
- {{CSSxRef("overflow-y")}}
- {{CSSxRef("overflow")}} shorthand
- {{CSSxRef("overflow-clip-margin")}}
- {{CSSxRef("scroll-behavior")}}
- {{CSSxRef("scrollbar-gutter")}}
- {{CSSxRef("text-overflow")}}

> - {{CSSxRef("block-ellipsis")}} (formerly {{CSSxRef("block-overflow")}}) {{experimental_inline}}

Reviewer 3: [MDN data](https://github.com/mdn/data/blob/main/css/properties.json#L2336-L2350) needs update.

- {{CSSxRef("continue")}} {{experimental_inline}}
- {{CSSxRef("line-clamp")}} {{experimental_inline}}
- {{CSSxRef("max-lines")}} {{experimental_inline}}

### Non-standard properties

> - {{CSSxRef("-webkit-line-clamp")}} {{non-standard_inline}}

Reviewer 3: This property is standardized in [Level 4](https://w3c.github.io/csswg-drafts/css-overflow-4/#webkit-line-clamp).

- {{cssxref("scroll-margin")}} CSS shorthand property
- {{cssxref("scroll-padding")}} CSS shorthand property
- {{cssxref("scroll-snap-align")}} CSS property
- {{cssxref("scroll-snap-stop")}} CSS property
- {{cssxref("scroll-snap-type")}} CSS property
- {{CSSxRef("::-webkit-scrollbar")}} pseudo-element
- [`scrollbar`](/en-US/docs/Web/Accessibility/ARIA/Roles/scrollbar_role) ARIA role
- Element {{domxref("Element.scroll", "scroll()")}} method
- Element {{domxref("Element.scrollBy", "scrollBy()")}} method
- Element {{domxref("Element.scrollIntoView", "scrollIntoView()")}} method
- Element {{domxref("Element.scrollTo", "scrollTo()")}} method
- Element {{domxref("Element.scrollTop", "scrollTop")}} property
- Element {{domxref("Element.scrollLeft", "scrollLeft")}} property
- Element {{domxref("Element.scrollWidth", "scrollWidth")}} property
- Element {{domxref("Element.scrollHeight", "scrollHeight")}} property
- Document {{domxref("Document.scroll_event", "scroll")}} event

## Specifications

{{Specifications}}

## See also

- [CSS scrollbars styling module](/en-US/docs/Web/CSS/CSS_scrollbars)
- [CSS scroll snap module](/en-US/docs/Web/CSS/CSS_Scroll_Snap)

> - [CSS OM view module](/en-US/docs/Web/CSS/CSSOM_View)

Reviewer 3: The [spec](https://w3c.github.io/csswg-drafts/cssom-view/) names itself as "CSSOM View Module".
